File: src/tauri/main/kernel/upstream-failure.js
// @ts-check

/**
 * @typedef {{ code: string; category: string; endpoint: string | null; messageKey: string }} UpstreamFailureDetails
 */

/** @type {Readonly<Record<string, string>>} */
export const UPSTREAM_FAILURE_FALLBACKS = Object.freeze({
    'network.timeout': 'The request timed out before the target service responded.',
    'network.connect_failed': 'Could not connect to the target service. Your network, VPN, proxy, or endpoint address may be unavailable.',
    'network.proxy_failed': 'Could not connect through the configured proxy.',
    'network.dns_failed': 'Could not find the target service address.',
    'network.tls_failed': 'Could not establish a secure connection.',
    'network.body_interrupted': 'The response was interrupted while it was being read.',
    'network.request_failed': 'Network request failed.',
});

/**
 * @param {unknown} value
 * @returns {UpstreamFailureDetails | null}
 */
export function asUpstreamFailureDetails(value) {
    if (!value || typeof value !== 'object' || Array.isArray(value) || value instanceof Error) {
        return null;
    }

    const record = /** @type {Record<string, unknown>} */ (value);
    const code = typeof record.code === 'string' ? record.code.trim() : '';
    const category = typeof record.category === 'string' ? record.category.trim() : '';
    const messageKey = typeof record.message_key === 'string'
        ? record.message_key.trim()
        : (typeof record.messageKey === 'string' ? record.messageKey.trim() : '');

    if (!code || !category || !messageKey) {
        return null;
    }

    const endpoint = typeof record.endpoint === 'string' && record.endpoint.trim()
        ? record.endpoint.trim()
        : null;

    return { code, category, endpoint, messageKey };
}

/**
 * @param {unknown} value
 * @param {number} depth
 * @returns {UpstreamFailureDetails | null}
 */
export function findUpstreamFailureDetails(value, depth = 0) {
    if (depth > 4 || value === null || value === undefined) {
        return null;
    }

    const direct = asUpstreamFailureDetails(value);
    if (direct) {
        return direct;
    }

    if (value instanceof Error) {
        const error = /** @type {Record<string, unknown>} */ (/** @type {unknown} */ (value));
        for (const key of ['details', 'data', 'error']) {
            if (Object.prototype.hasOwnProperty.call(error, key)) {
                const nested = findUpstreamFailureDetails(error[key], depth + 1);
                if (nested) {
                    return nested;
                }
            }
        }

        return findUpstreamFailureDetails(error.cause, depth + 1);
    }

    if (Array.isArray(value)) {
        for (const item of value) {
            const nested = findUpstreamFailureDetails(item, depth + 1);
            if (nested) {
                return nested;
            }
        }
        return null;
    }

    if (typeof value !== 'object') {
        return null;
    }

    const record = /** @type {Record<string, unknown>} */ (value);
    const keys = Object.keys(record);
    if (keys.length === 1 && keys[0] === 'UpstreamFailure') {
        return findUpstreamFailureDetails(record.UpstreamFailure, depth + 1);
    }

    for (const key of ['details', 'error', 'cause', 'data']) {
        if (Object.prototype.hasOwnProperty.call(record, key)) {
            const nested = findUpstreamFailureDetails(record[key], depth + 1);
            if (nested) {
                return nested;
            }
        }
    }

    return null;
}

/**
 * @param {UpstreamFailureDetails | null} details
 */
export function upstreamFailureFallbackText(details) {
    if (!details) {
        return '';
    }

    const message = UPSTREAM_FAILURE_FALLBACKS[details.code] || 'Upstream request failed.';
    return details.endpoint ? `${message} (${details.endpoint})` : message;
}

File: src-tauri/crates/tt-domain/src/models/upstream_failure.rs
use std::fmt;

use serde::Serialize;

pub const UPSTREAM_FAILURE_CATEGORY_NETWORK: &str = "network";

pub const UPSTREAM_NETWORK_TIMEOUT: &str = "network.timeout";
pub const UPSTREAM_NETWORK_CONNECT_FAILED: &str = "network.connect_failed";
pub const UPSTREAM_NETWORK_PROXY_FAILED: &str = "network.proxy_failed";
pub const UPSTREAM_NETWORK_DNS_FAILED: &str = "network.dns_failed";
pub const UPSTREAM_NETWORK_TLS_FAILED: &str = "network.tls_failed";
pub const UPSTREAM_NETWORK_BODY_INTERRUPTED: &str = "network.body_interrupted";
pub const UPSTREAM_NETWORK_REQUEST_FAILED: &str = "network.request_failed";

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct UpstreamFailure {
    pub code: String,
    pub category: String,
    pub endpoint: Option<String>,
    pub message_key: String,
}

impl UpstreamFailure {
    pub fn network(
        code: impl Into<String>,
        endpoint: Option<String>,
        message_key: impl Into<String>,
    ) -> Self {
        Self {
            code: code.into(),
            category: UPSTREAM_FAILURE_CATEGORY_NETWORK.to_string(),
            endpoint,
            message_key: message_key.into(),
        }
    }

    pub fn fallback_message(&self) -> &'static str {
        match self.code.as_str() {
            UPSTREAM_NETWORK_TIMEOUT => {
                "The request timed out before the target service responded."
            }
            UPSTREAM_NETWORK_CONNECT_FAILED => "Could not connect to the target service.",
            UPSTREAM_NETWORK_PROXY_FAILED => "Could not connect through the configured proxy.",
            UPSTREAM_NETWORK_DNS_FAILED => "Could not find the target service address.",
            UPSTREAM_NETWORK_TLS_FAILED => "Could not establish a secure connection.",
            UPSTREAM_NETWORK_BODY_INTERRUPTED => {
                "The response was interrupted while it was being read."
            }
            UPSTREAM_NETWORK_REQUEST_FAILED => "Network request failed.",
            _ => "Upstream request failed.",
        }
    }
}

impl fmt::Display for UpstreamFailure {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.endpoint.as_deref() {
            Some(endpoint) if !endpoint.is_empty() => {
                write!(formatter, "{} ({endpoint})", self.fallback_message())
            }
            _ => formatter.write_str(self.fallback_message()),
        }
    }
}
